src/cart/crt_corpc.c

```diff
@@ -725,10 +725,14 @@ crt_corpc_reply_hdlr(const struct crt_cb_info *cb_info)
 	D_SPIN_UNLOCK(&parent_rpc_priv->crp_lock);
 
 	if (req_done) {
+		bool am_root;
+
 		RPC_ADDREF(parent_rpc_priv);
 		crt_corpc_complete(parent_rpc_priv);
 
-		if (co_ops && co_ops->co_post_reply)
+		am_root = (co_info->co_grp_priv->gp_self ==
```

Comment:
Should we call this callback for all (not only NONE root)? Maybe move all free in single place?

Reply:
the problem is that when call co_post_reply, in corpc code here it does not know whether or not user will still use the memory (for example, user possibly addref the rpc and use its output for some other purpose).

Reply:
on leaf node or middle node, user lost control once the reply_send is called. But at root node, user's behavior is more flexible.

Reply:
Hmm. So, anyway it will be better to change current proc and call free for all structures at RPC destroy time.

I'm going to work for this patch later. Now let's land this solution.

Reply:
@liuxuezhao how is memory then free-ed on the root node? or does not not get allocated at all in this case on the root node as it is no longer sending the response to anyone and as such wont have those buffers?

Reply:
@dmiter for corpc's root node, the output is not processed by proc func at all.

@frostedcmos on root node, it is freed by user's corpc's initiator side. you may refer dimitry's https://github.com/daos-stack/daos/pull/5008

Reply:
root node free in "DAOS-6808 pool: Fix cleanuping resources #5008" in src/mgmt/srv_pool.c

Reply:
@liuxuezhao if RPC is sent then proc functions are called. How it's possible to not use proc functions at all?

Reply:
on corpc root node, the bcast rpc is a "virtual" RPC, it just create a set of "real" RPC to its children. When forward to children it just copy the input struct to child real RPC, go get reply it just aggregate child's real rpc reply.

Reply:
you may refer crt_corpc_req_hdlr()

Reply:
@liuxuezhao Thanks! I will learn this logic in more details.